core/silkworm/chain/validity.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+   Copyright 2021 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "validity.hpp"
+
+#include <silkworm/trie/vector_root.hpp>
+
+#include "difficulty.hpp"
+
+namespace silkworm {
+
+static std::optional<BlockHeader> get_parent(const StateBuffer& state, const BlockHeader& header) {
+    return state.read_header(header.number - 1, header.parent_hash);
+}
+
+ValidationError validate_block_header(const BlockHeader& header, const StateBuffer& state, const ChainConfig& config) {
+    // TODO[Issue 144] Ethash PoW verification
+
+    std::optional<BlockHeader> parent{get_parent(state, header)};
+    if (!parent) {
+        return ValidationError::kUnknownParent;
+    }
+
+    bool parent_has_uncles{parent->ommers_hash != kEmptyListHash};
+    intx::uint256 difficulty{canonical_difficulty(header.number, header.timestamp, parent->difficulty,
+                                                  parent->timestamp, parent_has_uncles, config)};
+    if (difficulty != header.difficulty) {
+        return ValidationError::kWrongDifficulty;
+    }
+
+    if (header.gas_used > header.gas_limit) {
+        return ValidationError::kGasAboveLimit;
+    }
+
+    uint64_t diff{header.gas_limit > parent->gas_limit ? header.gas_limit - parent->gas_limit
```

Comment:
To avoid confusion amongs diff and difficulty I'd rename this as `gas_delta` or `difference`